apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Trace/TraceTest.cs

```diff
@@ -0,0 +1,420 @@
+using Google.Cloud.Diagnostics.Common.Tests;
+using Google.Cloud.Trace.V1;
+using Google.Protobuf.WellKnownTypes;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+using TraceProto = Google.Cloud.Trace.V1.Trace;
+
+namespace Google.Cloud.Diagnostics.AspNetCore.IntegrationTest.Trace
+{
+    public class TraceTest
+    {
+        /// <summary>Total time to spend sleeping when looking for traces.</summary>
+        private static readonly TimeSpan _timeout = TimeSpan.FromSeconds(10);
+
+        /// <summary>Time to sleep between checks for traces.</summary>
+        private static readonly TimeSpan _sleepInterval = TimeSpan.FromSeconds(2);
+
+        /// <summary>Project id to run the test on.</summary>
+        private readonly string _projectId;
+
+        /// <summary>Client to use to send RPCs.</summary>
+        private readonly TraceServiceClient _client;
+
+        public TraceTest()
+        {
+            _projectId = Utils.GetProjectIdFromEnvironment();
+            _client = TraceServiceClient.Create();
+        }
+
+        /// <summary>
+        /// Gets a trace that contains a span with the given name.
+        /// </summary>
+        /// <param name="expectTrace">True if the trace is expected to exist.  This is used
+        ///     to minimize RPC calls.</param>
+        private async Task<TraceProto> GetTrace(string spanName, Timestamp startTime, bool expectTrace = true)
+        {
+            TimeSpan totalSleepTime = TimeSpan.Zero;
+            while (totalSleepTime <= _timeout)
+            {
+                TimeSpan sleepTime = expectTrace ? _sleepInterval : _timeout;
+                totalSleepTime += sleepTime;
+                Thread.Sleep(sleepTime);
+
+                ListTracesRequest request = new ListTracesRequest
+                {
+                    ProjectId = _projectId,
+                    StartTime = startTime,
+                    View = ListTracesRequest.Types.ViewType.Complete
+                };
+                var traces = _client.ListTracesAsync(request);
+                TraceProto trace = await traces.FirstOrDefault(t => t.Spans.Any(s => s.Name.Equals(spanName)));
+                if (trace != null)
+                {
+                    return trace;
+                }
+            }
+            return null;
+        }
+
+        [Fact]
+        public async Task Trace()
+        {
+            string testId = Utils.GetTestId();
+            var startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+
+            var builder = new WebHostBuilder().UseStartup<TraceTestNoBufferHighQpsApplication>();
+            using (TestServer server = new TestServer(builder))
```

Comment:
Nit: I would tend to use `var` here. No need to duplicate the type name. But it's up to you.